### PR TITLE
Switch to ppxlib from ppx_deriving

### PR DESCRIPTION
**Pull request by jonludlam**
Signed-off-by: Jon Ludlam <jonathan.ludlam@citrix.com>

### CODE REVIEW

File: ppx/ppx_deriving_rpc.ml [hunk 1]
```diff
@@ -0,0 +1,424 @@
+open Base
+open Ppxlib
+open Ast_builder.Default
+open Common
+
+let argn = Printf.sprintf "a%d"
+
+let rpc_of str = "rpc_of_" ^ str
+let of_rpc str = str ^ "_of_rpc"
+let map_lident f = function | Lident x -> Lident (f x) | Ldot (path, lid) -> Ldot (path, f lid) | Lapply _ -> failwith "boo"
```
[the remark this answers is not in the document]
Comment (jonludlam) [reply ppx/ppx_deriving_rpc.ml]: oops...!

File: ppx/jbuild [hunk 2]
```diff
@@ -1,16 +1,10 @@
...
 (library
  ((name ppx_deriving_rpc)
   (public_name ppx_deriving_rpc)
   (synopsis "[@@deriving rpc]")
   (kind ppx_deriver)
-  (libraries (ppx_deriving.api
-              rpclib.core))
-  (preprocess (action (run ppxfind -legacy ppx_tools.metaquot --as-pp ${<})))
+  (ppx_runtime_libraries (rpclib))
+  (libraries (ppxlib))
```
[the remark this answers is not in the document]
Comment (jonludlam) [reply ppx/jbuild]: yep.

File: ppx/ppx_deriving_rpcty.ml [hunk 3]
```diff
@@ -1,387 +1,316 @@
...
 
 module Typ_of = struct
-  let wrap_runtime decls =
-    [%expr
-      let open! Rpc.Types in
-      [%e decls]]
 
-  let rec expr_of_typ typ =
+  let rec expr_of_typ ~loc typ =
     let expr =
```
[the remark this answers is not in the document]
Comment (jonludlam) [reply ppx/ppx_deriving_rpcty.ml]: yeah, we had some issues with that when trying to marshal an ADT containing a constructor `String` - it got confused with the `String` type in `Rpc.Types` in some code that @rwdrich was doing. I've tried to make it all more specific in an effort to fix that though I've not retried what he was doing. We could perhaps turn it into a unit test - @rwdrich could you cut'n'paste here the code that was failing?

File: ppx/common.ml [hunk 4]
```diff
@@ -0,0 +1,135 @@
+open Base
+open Ppxlib
+open Ast_builder.Default
+
```
[the remark this answers is not in the document]
Comment (jonludlam) [reply ppx/common.ml]: Hmm, yes probably, Good idea.

File: ppx/ppx_deriving_rpcty.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (rwdrich) [reply ppx/ppx_deriving_rpcty.ml]: The code that was failing was in trying to add deriving rpcty to Schema.Value
` module Value = struct
   type t =
    | String of string
    | Set of string list
    | Pairs of (string * string) list
  [@@deriving sexp]`

My understanding of the problem was in trying to use rpcty with Schema.Value, the auto generated code has multiple instances of String. 
`let open! Rpc.Types` in is called above each of the type options, and as such inside the definition of String of string, the compiler doesn't know gets the wrong String. This means that it's trying to apply the rpc String, where it should be applying the Schema.Value String

File: ppx/ppx_deriving_rpc.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonludlam) [reply ppx/ppx_deriving_rpc.ml]: Fixed.

File: ppx/ppx_deriving_rpcty.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonludlam) [reply ppx/ppx_deriving_rpcty.ml]: I've added a unit test for this now (it passes, hurrah!)

**Comment by mseri** [comment]
You need to update `ppx_deriving_rpc.opam` before merging (the ppx tests are failing right now)

**Comment by mseri** [comment]
Also, please rebase on master, I should have fixed Travis

**Comment by mseri** [comment]
It is failing with
```
#=== ERROR while compiling ppx_deriving_rpc.master ============================#
# context              2.0.1 | linux/x86_64 | ocaml-base-compiler.4.06.0 | pinned(file:///repo)
# path                 ~/.opam/ocaml-base-compiler.4.06.0/.opam-switch/build/ppx_deriving_rpc.master
# command              ~/.opam/ocaml-base-compiler.4.06.0/bin/jbuilder build -p ppx_deriving_rpc -j 127
# exit-code            1
# env-file             ~/.opam/log/ppx_deriving_rpc-287-1f4e71.env
# output-file          ~/.opam/log/ppx_deriving_rpc-287-1f4e71.out
### output ###
#       ocamlc ppx/.ppx_deriving_rpc.objs/ppx_deriving_rpc.{cmi,cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/ocaml-base-compiler.4.06.0/bin/ocamlc.opt -w -40 -g -bin-annot -I ppx/.ppx_deriving_rpc.objs -I ppx/.ppx_deriving_rpc.objs/.private -I /home/opam/.opam/ocaml-base-compiler.4.06.0/lib/base -I /home/opam/.opam/ocaml-base-compiler.4.06.0/lib/base/caml -I /home/opam/.opam/ocaml-base-compiler.4.06.0/lib/base/shadow_stdlib -I /home/opam/.opam/ocaml-base-compil[...]
# File "ppx/ppx_deriving_rpc.ml", line 106, characters 44-49:
# Error: This expression has type
#          Ppxlib.label Ppxlib.loc = Ppxlib.label Location.loc
#        but an expression was expected of type Base.String.t = string
```

**Pull request merged**